File: next_url_hook.js
async function* fetchAllPages(url, params = {}, headers = {}) {
  while (url) {
    const response = await fetch(url, { headers });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data = await response.json();
    yield data.results || [];
    url = data.next_url;
    // Clear params for subsequent calls if necessary
    params = {};
  }
}

// Usage with async iteration
(async () => {
  for await (const pageResults of fetchAllPages("https://api.example.com/items")) {
    pageResults.forEach(item => console.log(item));
  }
})();

File: next_url_hook.py
import requests

def fetch_all_pages(url, params=None, headers=None):
    while url:
        response = requests.get(url, params=params, headers=headers)
        response.raise_for_status()
        data = response.json()
        yield data.get("results", [])
        # Update the URL to the next page if it exists
        url = data.get("next_url")
        # Reset params after the first call if needed
        params = None

# Usage
for page_results in fetch_all_pages("https://api.example.com/items"):
    for item in page_results:
        print(item)

